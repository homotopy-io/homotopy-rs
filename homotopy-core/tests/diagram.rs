use homotopy_core::{
    examples,
    signature::{GeneratorInfo, Signature},
    Boundary, Generator,
};
use insta::assert_debug_snapshot;

#[test]
fn assoc_left() {
    assert_debug_snapshot!(examples::associator().1.source());
}

#[test]
fn assoc_right() {
    assert_debug_snapshot!(examples::associator().1.target());
}

// ensure diagram with (source, target)
//  |
//  m     |
// / \  â¤³ |
// u  |   |
//    |
// can be created (source and target are each 3D)
#[test]
fn monoid_with_unit_to_id() -> anyhow::Result<()> {
    let (mut sig, s) = examples::scalar();
    let x = sig
        .generator_info(Generator::new(0, 0))
        .unwrap()
        .diagram()
        .clone();

    let monoid = sig.add(s.attach(&s, Boundary::Target, &[])?, s.clone())?;
    let unit = sig.add(x.identity().identity(), s.clone())?;
    let monoid_with_unit = monoid.attach(&unit, Boundary::Source, &[])?;
    assert!(sig.add(monoid_with_unit, s.identity()).is_ok());

    Ok(())
}
