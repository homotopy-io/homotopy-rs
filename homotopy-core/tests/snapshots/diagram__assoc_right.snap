---
source: homotopy-core/tests/diagram.rs
expression: "examples::associator().1.target()"
---
DiagramN {
    source: DiagramN {
        source: Diagram0(
            0:0⁺,
        ),
        cospans: [
            Cospan {
                forward: Rewrite0(
                    0:0⁺,
                    1:1⁺,
                    Label(
                        Some(
                            (1, BoundaryPath(Source, 0), []),
                        ),
                    ),
                ),
                backward: Rewrite0(
                    0:0⁺,
                    1:1⁺,
                    Label(
                        Some(
                            (1, BoundaryPath(Target, 0), []),
                        ),
                    ),
                ),
            },
            Cospan {
                forward: Rewrite0(
                    0:0⁺,
                    1:1⁺,
                    Label(
                        Some(
                            (1, BoundaryPath(Source, 0), []),
                        ),
                    ),
                ),
                backward: Rewrite0(
                    0:0⁺,
                    1:1⁺,
                    Label(
                        Some(
                            (1, BoundaryPath(Target, 0), []),
                        ),
                    ),
                ),
            },
            Cospan {
                forward: Rewrite0(
                    0:0⁺,
                    1:1⁺,
                    Label(
                        Some(
                            (1, BoundaryPath(Source, 0), []),
                        ),
                    ),
                ),
                backward: Rewrite0(
                    0:0⁺,
                    1:1⁺,
                    Label(
                        Some(
                            (1, BoundaryPath(Target, 0), []),
                        ),
                    ),
                ),
            },
        ],
    },
    cospans: [
        Cospan {
            forward: RewriteN {
                dimension: 1,
                cones: [
                    Cone {
                        index: 1,
                        internal: ConeInternal {
                            source: [
                                Cospan {
                                    forward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Source, 0), []),
                                            ),
                                        ),
                                    ),
                                    backward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Target, 0), []),
                                            ),
                                        ),
                                    ),
                                },
                                Cospan {
                                    forward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Source, 0), []),
                                            ),
                                        ),
                                    ),
                                    backward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Target, 0), []),
                                            ),
                                        ),
                                    ),
                                },
                            ],
                            target: Cospan {
                                forward: Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 1), []),
                                        ),
                                    ),
                                ),
                                backward: Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Target, 1), []),
                                        ),
                                    ),
                                ),
                            },
                            regular_slices: [
                                Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 1), []),
                                        ),
                                    ),
                                ),
                                Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 0), [Regular(1)]),
                                        ),
                                    ),
                                ),
                                Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Target, 1), []),
                                        ),
                                    ),
                                ),
                            ],
                            singular_slices: [
                                Rewrite0(
                                    1:1⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 0), [Singular(0)]),
                                        ),
                                    ),
                                ),
                                Rewrite0(
                                    1:1⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 0), [Singular(1)]),
                                        ),
                                    ),
                                ),
                            ],
                        },
                    },
                ],
            },
            backward: RewriteN {
                dimension: 1,
                cones: [
                    Cone {
                        index: 1,
                        internal: ConeInternal {
                            source: [
                                Cospan {
                                    forward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Source, 0), []),
                                            ),
                                        ),
                                    ),
                                    backward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Target, 0), []),
                                            ),
                                        ),
                                    ),
                                },
                            ],
                            target: Cospan {
                                forward: Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 1), []),
                                        ),
                                    ),
                                ),
                                backward: Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Target, 1), []),
                                        ),
                                    ),
                                ),
                            },
                            regular_slices: [
                                Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 1), []),
                                        ),
                                    ),
                                ),
                                Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Target, 1), []),
                                        ),
                                    ),
                                ),
                            ],
                            singular_slices: [
                                Rewrite0(
                                    1:1⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Target, 0), [Singular(0)]),
                                        ),
                                    ),
                                ),
                            ],
                        },
                    },
                ],
            },
        },
        Cospan {
            forward: RewriteN {
                dimension: 1,
                cones: [
                    Cone {
                        index: 0,
                        internal: ConeInternal {
                            source: [
                                Cospan {
                                    forward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Source, 0), []),
                                            ),
                                        ),
                                    ),
                                    backward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Target, 0), []),
                                            ),
                                        ),
                                    ),
                                },
                                Cospan {
                                    forward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Source, 0), []),
                                            ),
                                        ),
                                    ),
                                    backward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Target, 0), []),
                                            ),
                                        ),
                                    ),
                                },
                            ],
                            target: Cospan {
                                forward: Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 1), []),
                                        ),
                                    ),
                                ),
                                backward: Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Target, 1), []),
                                        ),
                                    ),
                                ),
                            },
                            regular_slices: [
                                Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 1), []),
                                        ),
                                    ),
                                ),
                                Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 0), [Regular(1)]),
                                        ),
                                    ),
                                ),
                                Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Target, 1), []),
                                        ),
                                    ),
                                ),
                            ],
                            singular_slices: [
                                Rewrite0(
                                    1:1⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 0), [Singular(0)]),
                                        ),
                                    ),
                                ),
                                Rewrite0(
                                    1:1⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 0), [Singular(1)]),
                                        ),
                                    ),
                                ),
                            ],
                        },
                    },
                ],
            },
            backward: RewriteN {
                dimension: 1,
                cones: [
                    Cone {
                        index: 0,
                        internal: ConeInternal {
                            source: [
                                Cospan {
                                    forward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Source, 0), []),
                                            ),
                                        ),
                                    ),
                                    backward: Rewrite0(
                                        0:0⁺,
                                        1:1⁺,
                                        Label(
                                            Some(
                                                (1, BoundaryPath(Target, 0), []),
                                            ),
                                        ),
                                    ),
                                },
                            ],
                            target: Cospan {
                                forward: Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 1), []),
                                        ),
                                    ),
                                ),
                                backward: Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Target, 1), []),
                                        ),
                                    ),
                                ),
                            },
                            regular_slices: [
                                Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Source, 1), []),
                                        ),
                                    ),
                                ),
                                Rewrite0(
                                    0:0⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Target, 1), []),
                                        ),
                                    ),
                                ),
                            ],
                            singular_slices: [
                                Rewrite0(
                                    1:1⁺,
                                    2:2⁺,
                                    Label(
                                        Some(
                                            (2, BoundaryPath(Target, 0), [Singular(0)]),
                                        ),
                                    ),
                                ),
                            ],
                        },
                    },
                ],
            },
        },
    ],
}
